apps/src/lab2/views/LabViewsRenderer.tsx

```diff
@@ -38,27 +41,60 @@ const LabViewsRenderer: React.FunctionComponent = () => {
   );
 
   const isViewingExemplar = getAppOptionsViewingExemplar();
-
-  const capitalizedLessonBackground = useAppSelector(
-    state =>
-      capitalizeFirstLetter(
-        getCurrentLesson(state)?.background || 'light'
-      ) as Theme
+  const lesson = useAppSelector(state => getCurrentLesson(state));
+  const {signInState} = useAppSelector(state => state.currentUser);
+  const dispatch = useAppDispatch();
+
+  // We only use the global user preference for theme if the current lesson has
+  // at least one python lab level or the current level is a python lab level.
+  const useThemeUserPreference = useMemo(
+    () =>
+      levelProperties?.appName === 'pythonlab' ||
+      lesson?.levels.some((level: Level) => level.app === 'pythonlab'),
+    [lesson?.levels, levelProperties?.appName]
   );
 
   // Set the theme for the current app.
   const {setTheme} = useTheme();
   useEffect(() => {
     if (currentAppName) {
       const supportedThemes = lab2EntryPoints[currentAppName]?.themes;
-
-      if (supportedThemes.includes(capitalizedLessonBackground)) {
-        setTheme(capitalizedLessonBackground);
+      dispatch(setIsLoadingTheme(true));
+
+      const setThemeHelper = () => {
+        // If the body has a class use that to set the theme, if its supported by the lab.
+        // Otherwise, use the first supported theme.
+        // We will only run this if we are not using the user preference, so it's safe to pull
+        // from the body class, as the user is not dynamically changing the theme.
+        const bodyClassList = document.body.classList;
+        const bodyTheme = bodyClassList.contains('background-light')
+          ? 'Light'
+          : 'Dark';
+        if (supportedThemes.includes(bodyTheme)) {
+          setTheme(bodyTheme);
+        } else {
+          setTheme(supportedThemes[0]);
+        }
+        dispatch(setIsLoadingTheme(false));
+      };
+
+      if (useThemeUserPreference && signInState === SignInState.SignedIn) {
+        const fetchAndSetTheme = async () => {
```

Comment:
nit: could pull the `useEffect` code out into its own helper in order to call `async`/`await` directly instead of having to create inner functions?

Reply:
Can you explain what you mean here?

Reply:
Something like:

```
const updateTheme = async () => {
   const supportedThemes = lab2EntryPoints[currentAppName]?.themes;
   dispatch(setIsLoadingTheme(true));
   
   let userTheme = getBodyTheme(); // could pull out another helper that reads body class/possibly share with similar code in Lab2Wrapper
   
   if (useThemeUserPreference && signInstate === SignInState.SignedIn) {
      userTheme = await new UserPreferences().getGlobalTheme();
   }

   if (userTheme && supportedThemes.includes(userTheme)) {
      setTheme(userTheme);
   } else {
      setTheme(supportedThemes[0]);
   }

   dispatch(setIsLoadingTheme(false));
}

// then in useEffect

useEffect(() => {
   if (currentAppName) {
      updateTheme();
   }
}, [currentAppName]);
```